
libstdc++: replace deprecated std::is_pod<T> in blink

std::is_pod is deprecated since C++20. Replace with std::trivial and
std::is_standard_layout. Avoids a lot of warnings.

Bug: 957519
Change-Id: Ia5ffac10b68420eb68eb34b2eeb8afb574c039a6
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4936811
Reviewed-by: Michael Lippautz <mlippautz@chromium.org>
Commit-Queue: Stephan Hartmann <stha09@googlemail.com>
Cr-Commit-Position: refs/heads/main@{#1209300}

Updated with fix from 
https://chromium.googlesource.com/chromium/src/+/7e7f280e2438acfe1bade26dad25a808885239a3%5E%21/#F0

--- a/third_party/blink/renderer/platform/wtf/hash_traits.h
+++ b/third_party/blink/renderer/platform/wtf/hash_traits.h

@@ -194,7 +194,8 @@
   struct NeedsToForbidGCOnMove {
     // TODO(yutak): Consider using of std:::is_trivially_move_constructible
     // when it is accessible.
-    static constexpr bool value = !std::is_pod<T>::value;
+    static constexpr bool value =
+        !std::is_trivial_v<T> || !std::is_standard_layout_v<T>;
   };
 
   // The kCanTraceConcurrently value is used by Oilpan concurrent marking. Only

--- a/third_party/blink/renderer/platform/wtf/text/string_hasher.h
+++ b/third_party/blink/renderer/platform/wtf/text/string_hasher.h

@@ -191,7 +191,8 @@
   void AddCharactersAssumingAligned_internal(const unsigned char* data, unsigned length) {
     DCHECK(!has_pending_character_);
 
-    static_assert(std::is_pod<T>::value, "we only support hashing POD types");
+    static_assert(std::is_trivial_v<T> && std::is_standard_layout_v<T>,
+                  "we only support hashing POD types");
     bool remainder = length & 1;
     length >>= 1;
 
@@ -216,7 +217,8 @@
 
   template <typename T, UChar Converter(T)>
   void AddCharacters_internal(const unsigned char* data, unsigned length) {
-    static_assert(std::is_pod<T>::value, "we only support hashing POD types");
+    static_assert(std::is_trivial_v<T> && std::is_standard_layout_v<T>,
+                  "we only support hashing POD types");
 
     if (has_pending_character_ && length) {
       has_pending_character_ = false;

